### PR TITLE
[DRAFT][SDP][BigPR] Base all sync on torch futures

### PR DESCRIPTION
**Pull request by blefaudeux**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?
Fixes # (issue).

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

@girifb

### CODE REVIEW

File: fairscale/nn/data_parallel/sharded_ddp.py [hunk 1]
```diff
@@ -416,35 +403,28 @@ def reduce(*_: Any) -> None:
 
                     # Make sure that this is not fired twice
                     self._grad_to_be_reduced[index] = False
-                    param.grad.mul_(self.world_size_scaling)
 
-                    if self.reduce_fp16:
-                        param.grad.data = param.grad.data.half()
+                    param.grad.mul_(self.world_size_scaling)
```
Comment (msbaines) [inline fairscale/nn/data_parallel/sharded_ddp.py:407]: What happened to the` if self.reduce_fp16`

File: fairscale/optim/oss.py [hunk 2]
```diff
@@ -292,10 +294,20 @@ def clip_grad_norm(
...
+        assert not torch.isnan(clip_coef), (
+            "Clipping grads with a NaN, cannot be right.\n"
+            + f"Local norm: {local_norm.item()}. Total norm: {total_norm.item()}"
+        )
+
         if clip_coef < 1:
-            for device, device_params in self.per_device_params.items():
-                for p in filter(lambda x: x.grad is not None, device_params[self.rank]):
-                    p.grad.detach().mul_(clip_coef.to(device))  # type: ignore
+            with torch.no_grad():
```
Comment (msbaines) [inline fairscale/optim/oss.py:307]: This seems like unrelated cleanup. With large changes it is much easier to review if there is a single thesis.

File: fairscale/nn/data_parallel/sharded_ddp.py (hunk 1 above)
Comment (blefaudeux) [reply fairscale/nn/data_parallel/sharded_ddp.py:407]: it did not give the correct results, I did not test it well enough. It can be re-implemented on top of this PR, I was genuinely trying to keep things "small"

File: fairscale/optim/oss.py (hunk 2 above)
Comment (blefaudeux) [reply fairscale/optim/oss.py:307]: you're right, I'll move that to another PR

**Comment by blefaudeux** [comment]
I was planning to write down some explanations @msbaines, that's why it was in draft. Writing some history here: basically this started with a report from @zhengwy888 that he was seeing a different grad norm with shardedDDP when the buckets were activated (which led to https://github.com/facebookresearch/fairscale/issues/517).

I could repro that in a new unit test, which is part of this PR, turns out that the synchronization in between compute and comms was not correct, and shardedDDP did in fact produce different results from DDP. I really tried to get things to work without the torch futures, but the TL; DR is that you don't get to synchronize with `torch.distributed.reduce()` for instance, because this goes to a "hidden" cuda stream (a pool actually). The handles that `torch.distributed` returns are not a good synchronization primitive, they just mark the op being inserted in the cuda stream, not completeness. So the only reasonable way that I could find, which prove reliable, was to use `torch.futures`, which implied rewriting a fair bunch of things. Gloo does not support torch.futures, hence the trail of changes in the unit test, other than the new ones the other ones are unchanged (just backend limited to NCCL).

If you can think of another method I would love to learn about it. cc @mrshenli just in case I writing something wrong in there

File: fairscale/optim/oss.py [hunk 3]
```diff
@@ -245,6 +245,7 @@ def clip_grad_norm(
         max_norm: Union[float, int],
         norm_type: Union[float, int] = 2.0,
         filter_params_fn: Callable[[Any], Any] = None,
+        skip_nan_grads: bool = False,
```
Comment (blefaudeux) [inline fairscale/optim/oss.py:248]: will be moved to another PR, unrelated

File: tests/nn/data_parallel/test_sharded_ddp_features.py [hunk 4]
```diff
@@ -22,7 +22,6 @@
 from fairscale.utils.testing import (
```
Comment (blefaudeux) [inline tests/nn/data_parallel/test_sharded_ddp_features.py:22]: no changes in this file except from
- removing duplicated test with the other sharded_ddp test file (probably a botched merge)
- removing gloo tests

File: tests/nn/data_parallel/test_sharded_ddp_pytorch_parity.py [hunk 5]
```diff
@@ -201,28 +196,30 @@ def closure_sharded(input_tensor=input_tensor):
 @pytest.mark.parametrize("reduce_buffer_size", [0, 2 ** 20])
 @pytest.mark.parametrize("grad_accumulation", [True, False])
 @pytest.mark.parametrize("change_train_graph", [True, False])
-@pytest.mark.parametrize("fp16_reduction", _test_fp16_reduction)
-def test_ddp_parity(reduce_buffer_size, grad_accumulation, change_train_graph, fp16_reduction):
+@pytest.mark.parametrize("clip_grad_norm", [True, False])
```
Comment (blefaudeux) [inline tests/nn/data_parallel/test_sharded_ddp_pytorch_parity.py:199]: two new cases, which showed the problem that @zhengwy888 had spotted: check clip_grad_norm parity, with and without AMP, without and without buckets, with and without grad accumulation, and change in training graph. This reliably showed an error when the buckets were activated

Comment (blefaudeux) [reply tests/nn/data_parallel/test_sharded_ddp_pytorch_parity.py:199]: Screen Shot 2021-03-16 at 10 53 09 AM

File: tests/nn/data_parallel/test_sharded_ddp_pytorch_parity.py [hunk 6]
```diff
@@ -290,341 +285,9 @@ def run_ddp_parity_two_optim(rank, world_size, backend, temp_file_name, reduce_b
...
-    backend = dist.Backend.NCCL
     mp.spawn(
         run_ddp_parity_two_optim,
-        args=(world_size, backend, tempfile.mkstemp()[1], reduce_buffer_size),
-        nprocs=world_size,
-        join=True,
-    )
-
-
-def run_test_two_inputs(rank, world_size, backend, device, temp_file_name, reduce_buffer_size):
```
Comment (blefaudeux) [inline tests/nn/data_parallel/test_sharded_ddp_pytorch_parity.py:302]: all the test below were duplicates, probably a merge which failed on somebody. These are not ddp parity, they are test of sharded_ddp features and are handled in the other test file

File: fairscale/nn/data_parallel/sharded_ddp.py [hunk 7]
```diff
@@ -59,9 +58,6 @@ class ShardedDataParallel(nn.Module):
             (default: True) Check whether the parameters trainability (`requires_grad`) has changed and update both ShardedDDP
             and OSS automatically if this is the case. If set to False, `refresh_trainable()` needs to be called anytime
             a parameter is frozen or unfrozen.
-        reduce_fp16 (bool):
```
Comment (blefaudeux) [inline fairscale/nn/data_parallel/sharded_ddp.py:62]: this was broken and could be re-introduced later, not too complicated to do once the syncs are correct. I did try to keep this PR as small as possible

File: fairscale/nn/data_parallel/sharded_ddp.py [hunk 8]
```diff
@@ -317,18 +304,15 @@ def sync_buffers(self, blocking: bool = False) -> None:
             blocking (bool): wait for the operation to conclude.
         """
 
-        work_handles = []
+        futures = []
 
         for buffer in self.module.buffers(recurse=True):
-            work_handles.append(
```
Comment (blefaudeux) [inline fairscale/nn/data_parallel/sharded_ddp.py:323]: basically the crux of the issue is that, the handles are not a good sync primitive, they are just not correct, and the cuda events are not accessible because the stream being used is not accessible

File: fairscale/nn/data_parallel/sharded_ddp.py [hunk 9]
```diff
@@ -339,12 +323,18 @@ def zero_grad(self, set_to_none: bool = False) -> None:
                 See :meth:`torch.optim.Optimizer.zero_grad` for details.
         """
 
+        if self.auto_refresh_trainable:
```
Comment (blefaudeux) [inline fairscale/nn/data_parallel/sharded_ddp.py:326]: this is orthogonal but a bugfix: if the model trainability changed, and people zero_grad, then some of the gradients could skip being set to zero, which would break the subsequent grad norm computation..

File: fairscale/nn/data_parallel/sharded_ddp.py [hunk 10]
```diff
@@ -491,6 +454,10 @@ def _setup_backward_hooks(self) -> None:
         This makes the gradient reduction automatic whenever there's a backward pass
         """
 
+        # Empty the possible previous hooks
```
Comment (blefaudeux) [inline fairscale/nn/data_parallel/sharded_ddp.py:457]: suggested in the doc, doing that for correctness, I'm not sure of the effect. Remove the previous hooks before attaching new ones

File: fairscale/nn/data_parallel/sharded_ddp.py [hunk 11]
```diff
@@ -551,82 +515,68 @@ def _setup_bucket_strategy(self) -> None:
...
+        self._should_bucket_grad = []
+        self._should_bucket_grad = [False for _ in range(len(self._trainable_params))]
 
-        for param in self._trainable_params:
+        for index, param in enumerate(self._trainable_params):
             device = param.device
             dst_rank = self._trainable_param_to_rank[param]
 
             if param.device not in self.buckets.keys():
-                self.buckets[param.device] = [
```
Comment (blefaudeux) [inline fairscale/nn/data_parallel/sharded_ddp.py:560]: this could lead to a subtle over provision, if not all ranks where present for all devices. Moving that to allocating the bucket on demand, when it's really required

File: fairscale/nn/data_parallel/sharded_ddp.py [hunk 12]
```diff
@@ -551,82 +515,68 @@ def _setup_bucket_strategy(self) -> None:
...
+            bucket.shrink()
 
-            else:
-                self._should_bucket_grad.append(False)
+            # - store the callback which releases the bucket
+            def cleanup(fut: torch.futures.Future) -> None:
+                if bucket.destination != self.global_rank:
+                    fut.wait()
+                    # FIXME: Autograd dies with this
+                    # bucket.collapse()
```
Comment (blefaudeux) [inline fairscale/nn/data_parallel/sharded_ddp.py:555]: this is an extra, not done in the current (master) version of ShardedDDP: release the bucket when reduced, rebuild in the next FW. For some reason autograd segfaults on that, opinions welcome

File: fairscale/nn/data_parallel/sharded_ddp.py [hunk 13]
```diff
@@ -551,82 +515,68 @@ def _setup_bucket_strategy(self) -> None:
...
 
-        self._bucket_list = list(chain(*[self.buckets[device] for device in self.buckets.keys()]))
+            bucket.callback = cleanup
 
-        # Resize the buckets to remove lost space in the end
-        for bucket in self._bucket_list:
-            bucket.buffer.resize_(bucket.fill)
-            bucket.sent = True
-
-    def _consume_work_handles(self) -> None:
```
Comment (blefaudeux) [inline fairscale/nn/data_parallel/sharded_ddp.py:596]: all this was nice, but handle.wait() does not do the same thing as (torch.future).wait()... this was an expensive mistake

File: fairscale/nn/misc/grad_bucket.py [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+                self._add_grad_as_view(p)
+
+            self._is_collapsed = False
+
+    @torch.no_grad()
+    def scale_and_reduce(self, scale: float) -> torch.futures.Future:
+        assert self._buffer is not None, "Cannot send a collapsed bucket, please rebuild"
+
+        self._buffer.mul_(scale)
+        torch.cuda.synchronize()
```
Comment (blefaudeux) [inline fairscale/nn/misc/grad_bucket.py:111]: this is very expensive in terms of time, but the results are not correct without it. If a reviewer can think of a better synchronization that would be great..

File: fairscale/nn/misc/grad_bucket.py [hunk 15]
```diff
@@ -0,0 +1,148 @@
+from typing import Any, Callable, List, Optional, Union
+
```
Comment (blefaudeux) [inline fairscale/nn/misc/grad_bucket.py:2]: the idea behind that was to simplify sharded_ddp, which is already complicated enough. The reasoning is the same as before, just with the torch.futures and not workhandle

**Comment by blefaudeux** [comment]
(rebased and some unrelated changes removed, still not ready for review)

File: tests/nn/data_parallel/test_sharded_ddp_features.py (hunk 4 above)
Comment (blefaudeux) [reply tests/nn/data_parallel/test_sharded_ddp_features.py:22]: test removal has been moved to another PR, now just gloo backend removal as expected

File: fairscale/nn/data_parallel/sharded_ddp.py (hunk 1 above)
Comment (blefaudeux) [reply fairscale/nn/data_parallel/sharded_ddp.py:399]: resolved in the newer version

File: tests/nn/data_parallel/test_sharded_ddp_pytorch_parity.py (hunk 5 above)
Comment (blefaudeux) [reply tests/nn/data_parallel/test_sharded_ddp_pytorch_parity.py]: checked in with another PR

File: tests/nn/data_parallel/test_sharded_ddp_pytorch_parity.py (hunk 6 above)
Comment (blefaudeux) [reply tests/nn/data_parallel/test_sharded_ddp_pytorch_parity.py]: checked in in another PR

**Comment by blefaudeux** [comment]
This kills compatibility with torch 1.6 it seems, looks like the torch.distributed calls are not torch.futures compatible :(

**Comment by blefaudeux** [comment]
not supported on torch 1.6, will need to wait for 1.9 so that we deprecate it
